src/utils/BlockUtils.ts

```diff
@@ -166,3 +169,40 @@ export class BlockFinder {
     }
   }
 }
+
+/**
+ * @notice Get the block number for a given timestamp fresh from on-chain data if not found in redis cache.
+ * If redis cache is not available, then requests block from blockFinder.
+ * @param chainId Chain to load block finder for.
+ * @param timestamp Approximate timestamp of the to requested block number.
+ * @param blockFinder Caller can optionally pass in a block finder object to use instead of creating a new one
+ * or loading from cache. This is useful for testing primarily.
+ * @returns Block number for the requested timestamp.
+ */
+export async function getCachedBlockForTimestamp(
+  chainId: number,
+  timestamp: number,
+  blockFinder: BlockFinder,
+  provider: Provider,
+  cache?: CachingMechanismInterface
+): Promise<number> {
+  // If no redis client, then request block from blockFinder. Otherwise try to load from redis cache.
+  if (!isDefined(cache)) {
+    return (await blockFinder.getBlockForTimestamp(timestamp)).number;
+  }
+  const key = `${chainId}_block_number_${timestamp}`;
```

Comment:
yeah so we're going to have to break up the Promise.all into two calls:
- first get current block.
- Check `shouldCache`.
- If shouldCache, try to load from cache, etc.
- otherwise, load fresh

Reply:
Why do we need current block? Can we just compute the local (js) timestamp? Wouldn't that be good enough?

Reply:
https://github.com/across-protocol/sdk-v2/pull/438/commits/2015da59e59943f7babf309bc60ceab5601fe315